org.eclipse.xtext/src/org/eclipse/xtext/validation/DefaultUniqueNameContext.java

```diff
@@ -0,0 +1,317 @@
+
+		@Inject
+		private IContainer.Manager containerManager;
+		
+		@Override
+		public Context tryGetContext(Resource resource, CancelIndicator cancelIndicator) {
+			IResourceDescriptions index = getIndex(resource);
+			if (index == null) {
+				return null;
+			}
+			IResourceDescription description = getResourceDescription(resource);
+			if (description == null) {
+				return null;
+			}
+			IContainer container = containerManager.getContainer(description, index);
+			return new DefaultUniqueNameContext(description, container, getCaseInsensitivityHelper(), cancelIndicator);
+		}
+		
+		@Override
+		public boolean isAffected(
+				Collection<IResourceDescription.Delta> deltas,
+				IResourceDescription candidate,
+				IResourceDescriptions context) {
+			IContainer container = containerManager.getContainer(candidate, context);
+			for(IResourceDescription.Delta delta: deltas) {
+				if (delta.getNew() == null) {
+					if (intersects(delta.getOld(), candidate, true)) {
+						return true;
+					}
+				} else {
+					if (container.getResourceDescription(delta.getUri())!=null) {
+						if (isAffected(delta, candidate, true)) {
+							return true;
+						}
+					}
+				}
+			}
+			return false;
+		}
+
+	}
+	
+	/**
+	 * Provide a context for all visible {@link IContainer containers}.
+	 */
+	@Singleton
+	public static class VisibleContainers extends BaseGlobalContextProvider {
+
+		public static class Selectable extends AbstractCompoundSelectable {
+
+			private final List<IContainer> containers;
+
+			public Selectable(List<IContainer> containers) {
+				this.containers = containers;
+			}
+
+			@Override
+			protected Iterable<? extends ISelectable> getSelectables() {
+				return containers;
+			}
+			
+		}
+		
+		@Inject
+		private IContainer.Manager containerManager;
+		
+		@Override
+		public Context tryGetContext(Resource resource, CancelIndicator cancelIndicator) {
+			IResourceDescriptions index = getIndex(resource);
+			if (index == null) {
+				return null;
+			}
+			IResourceDescription description = getResourceDescription(resource);
+			if (description == null) {
+				return null;
+			}
+			List<IContainer> containers = containerManager.getVisibleContainers(description, index);
+			return new DefaultUniqueNameContext(description, new Selectable(containers), getCaseInsensitivityHelper(), cancelIndicator);
+		}
+		
+		
+		
+		@Override
+		public boolean isAffected(Collection<IResourceDescription.Delta> deltas, IResourceDescription candidate,
+				IResourceDescriptions context) {
+			List<IContainer> containers = containerManager.getVisibleContainers(candidate, context);
+			for(IResourceDescription.Delta delta: deltas) {
+				if (delta.getNew() == null) {
+					if (intersects(delta.getOld(), candidate, true)) {
+						return true;
+					}
+				} else {
+					for(IContainer container: containers) {
+						if (container.getResourceDescription(delta.getUri())!=null) {
+							if (isAffected(delta, candidate, true)) {
```

Comment:
does this work on renames?

Reply:
You mean file renames? Those should provide two deltas.

Reply:
hmmm

Reply:
Marking as resolved.